### DISCUSSION TITLE
PyScript compatibility

### DISCUSSION THREAD

**Discussion by chrisbuerginrogers** [petercorke/robotics-toolbox-python#458] [Ideas]
Hi all - thanks for the robotics toolbox (and @Peter - sweet website - I have used it a bit in the past) - looks pretty awesome - I am hoping to embed it into a [PyScript](https://pyscript.com) page so that I can customize (and then my students can too) a UI for robot control.  It would be great to embed Swift as well - if possible.  This is for an [introductory robotics course](https://shimmering-cesium-c35.notion.site/ME-35-2024-The-Playground-8af7d3371fb340bfa0817a85b9a2c1de) at Tufts University.  If you all are at all interested, let me introduce Nicholas (@ntoll ) - he is part of the PyScript team at Anaconda (he and I do a lot of education robotics stuff - he, Mu and I, LEGO).  If you all are interested - it would be cool to also link it with a bunch of ROS/Jupyter Lab stuff we have running as well.  Thanks

**Comment by ntoll** [comment]
:wave: Hi folks... Thanks for the ping Chris. 

Umm... so @chrisbuerginrogers is a totally amazing professor at Tufts, Boston, USA. I'm based in the UK and work on the open source project [PyScript](https://pyscript.net), which is basically a platform for Python in the browser (using WASM). We've been collaborating for, well I can't remember how long, so it must be a long time... and we both share a passion for engineering education. 

The thing about PyScript and Python in the browser is that you don't have to install anything on your local machine (it's all in the browser, shared via a URL), Python packages you may use are kept within the browser sandbox, AND you get access to all the funky APIs modern browser have, such as BLE, web-serial, camera, speech synthesis and all sorts of fun stuff. Clearly Chris sees your project as something to explore for pedagogical purposes, and wouldn't it be cool to be able to use it with just a modern web browser..? 

How can we help make this happen? (If this is of interest.)

**Comment by traversaro** [comment]
fyi @tobias-fischer, given the work presented in https://arxiv.org/pdf/2409.09941 .

**Comment by Tobias-Fischer** [comment]
Hi @chrisbuerginrogers and @ntoll,
This sounds like a great idea - I fully agree that embedding the toolbox+Swift would make students' lives a lot better! I am now teaching Peter's Introduction to Robotics unit at QUT so I have a significant interest in engineering education as well :).

As @traversaro said, we've been working toward a very similar aim, and you can read what we've been up to in https://arxiv.org/abs/2409.09941. Rather than using PyScript, we've been opting for https://emscripten-forge.org/. In addition to Peter's toolbox, we've also managed to compile ROS2 for WASM (which was much more challenging ..).

At the moment, the demos are scattered and not yet polished:
- https://ros2wasm.dev/ (ROS2 general demos)
- https://mbatc.github.io/swift-wasm-test/ (Swift)
- https://mbatc.github.io/wasm-cpp-test/lab/ (rclpy)
All the code for these is available as well, which might help in your PyScript endeavours. The robotics toolbox itself was straightforward to compile for WASM, Swift needed some changes, and ROS2 significant changes ..

Note that I got some funding so that @nmarticorena will work on polishing these demos in a few weeks time. Would be great to collaborate if you are interested.

**Reply by chrisbuerginrogers** [reply]
Excellent - I will follow up with an email around collaboration.  I tried out your ROS2 WASM site - very cool - @ntoll - can we leverage that?  Having ROS2 in a PyScript page would be excellent.  I was challenged by iRobot to see if we could have first years using ROS2 with their Create - [this](https://github.com/ceeoinnovations/EN1-iRobot) was the result.  It worked - but required a Tufts server with ROS2 installed - your solution would be more elegant.

**Comment by ntoll** [comment]
Hi Folks :wave: 

This is such an interesting endeavour for all sorts of reasons: Python, WASM, conda (my employer is Anaconda, originators of this technology, although not the same as emscriptenforge).

So I guess we should compare notes.

[PyScript](https://pyscript.net) ([technical docs here](https://docs.pyscript.net), [source code here](https://github.com/pyscript/pyscript)) is simply an easy to use platform for Python in the browser. With just a single reference in the `<head>` of your HTML document a whole bunch of magic happens so it's easy to write Python, define an environment, use web workers, access the browser's APIs and even select the most appropriate interpreter (MicroPython and CPython (Pyodide) are supported). Our focus is to make it as easy as possible to use the combined power of Python and the browser, in a flourishing and friendly open source community (we [hang out on discord](https://discord.gg/HxvBtukrg2)).

We chose Pyodide as our version of CPython because it is simply the most mature project, we have an excellent collaborative relationship with its core maintainers, they have already working versions of the most commom C-based tools such as `numpy`, `pandas`,  `PIL`, `matplotlib` etc..., and it's "standard" CPython (relatively unopinionated about things). I'll come to MicroPython in a moment... ;-)

I've very quickly skimmed the paper to which you link. As far as I can tell (caveat - I've probably missed something), the core surrounds making `rmw-wasm-cpp`, `wasm-cpp`, and `wasm-js` come to life. Although I've not explored the details, I guess this is wrapped in a Python API and I think you package all this up via a conda recipe? While I'm hedging my bets here, to make this work with Pyodide, "all one need do" (note the deliberate air-quotes there) is package your work as a C-extension compiled to WASM and available via PyPI -- i.e. it's "just" normal vanilla Python packaging. Pyodide [document this process](https://pyodide.org/en/stable/development/new-packages.html) and folks have even started building tooling to make it easier (e.g. [exodide](https://github.com/ymd-h/exodide) - although I'm not sure how "alive" that project currently is).

Regarding MicroPython, it's very small, fast and works exceptionally well in a browser context. When you don't need everything-but-the-kitchen-sink in Pyodide, but just a lightweight Python, MicroPython is ideal. Furthermore, there's a well documented way to create Python APIs that wrap C code, in MicroPython. This is *very* different to the world of CPython, most particularly because the C code is compiled into a bespoke build of MicroPython itself. That is delivered to the browser as a single asset. I'm mentioning this, not only for size/speed or startup reasons, but because Damien (creator of MicroPython) is a former Robo World Cup participant (so has an advanced background in robotics) and is Australian. ;-) I'm just thinking of the Venn diagram of shared interests and how such overlaps often lead to fruitful collaborations. Also, Chris has made extensive use of MicroPython in embedded (robotic) devices controlled by various microcontrollers... so the "stack" is always the familiar MicroPython. See, especially, is amazing work (and the work of his students) with Lego Spike Prime devices.

Anyway. How does this sound..? Happy to answer questions and dig into details further. If you're happy for me to do so, I can bring in various PyScript collaborators and some `conda` folks too - I think they'd love to learn about your use cases and experience of conda in a WASM context.

**Reply by Tobias-Fischer** [reply]
Hi @ntoll - let’s loop in @DerThorsten / @IsabelParedes / @wolfv who are much more informed about differences and similarities between Pyodide and emscripten-forge / PyScript and PyJS.

What advantages do you see of using Pyodide over emscripten-forge? We have NumPy etc there as well. Is it mainly the MicroPython aspect?

**Reply by Tobias-Fischer** [reply]
We should probably also move this discussion elsewhere by the way :)

**Reply by ntoll** [reply]
Hey hey.

The advantages? Chris gets his cake and eats it. ;-)

My motivation (and my default modus operandi) is to discover interesting and innovative ways in which folks can complement and amplify each other's work while mutually learning, growing and collaborating. Open source is my thing, creating a space in which communities can flourish is generally my approach. In my experience, when this way of working is embraced unforeseen yet rather wonderful outcomes tend to happen. This feels like a good opportunity for such a process. Also, ROS becomes useful in several web based contexts (be that PyScript, Pyodide, MicroPython or emscripten-forge contexts). What's not to like about that? ;-)

Also, yes about discussion location... I'm flexible, so please suggest alternatives (although I appreciate that discussions in GitHub are in the open for everyone to follow and find).

**Reply by chrisbuerginrogers** [reply]
Hi all - I am happy to move the discussion anywhere - but my "cake" is that I want students to be able to focus on key concepts, not installations (and we all know the headaches there - with large range of student computers).  My dream is Notion+Jupyter+PyScript+GitHub+ROS2 in one tight package (it should be as easy to drop in some ROS calls and run them as it is to drop in a mathematical equation).  Last year I brought openCV into my class with PyScript and within 45 min all students had something that took an image, subtracted different planes, looked at histograms, and found blobs.  This year in 30 min, they had TeachableMachines controlling their robots.  Note in all cases there was no install.  The power of Notion is they can then reflect on what they found (with multimedia).  The advantage of GitHub is that their code is properly stored and updated.  Jupyter and PyScript are similar - but I cannot get Jupyter to talk nicely to hardware over USB without a lot of secondary installs.  This is not an issue in PyScript.   Also in PyScript I can teach about user interface. [ This site](https://chrisrogers.pyscriptapps.com/cbr-libraries/latest/) has all the different things I use.  Note that you can try it all out with one click (maybe this is old news to you all but it is great to me - after 30+ years of working in K-16 classrooms).  Thanks

**Reply by fpliger** [reply]
Interesting conversation! @jezdez @wolfv and myself talked about this (or some very similar flavor) earlier this year and we've talked about this in the past. I'd love to resume this and collaborate towards shared goals.

**tldr;** what I'd like to see happening is the decoupling `emscripten-forge` packaging from the python interpreter itself (and maybe it's already the case in terms of design, idk), so that an user can use `emscripten-forge` to manage their dependencies/environments and can choose what underlying CPython interpreter it's using. I think this would benefit the community at large, positively influencing everyone in the ecosystem. With that said, I don't think any of it is a blocker to support `emscripten-forge`

**Longer Version** 

PyScript is a pluggable platform and, fwiw, aims support as many interpreters as are needed by our users. So, from the PyScript project perspective, in on what interpreter works best for what scenario (ex: need something light, fast and don't need specific CPython libraries? MicroPython is likely the best candidate for loading perf reasons! Need to use specific CPython dependencies? A CPython compatible interpreter (Pyodide or what is in `emscrpten-forge` are better choices!). 

We can definitely look into `emscripten-forge` being a supported platform in PyScript. I think it'd greatly benefit the entire community interested in Python and WASM (PyScript and `emscripten-forge` included). In addition to that, I do think that it'd also benefit the entire community if `emscripten-forge` were interpreter agnostic (in ways) and allowed users to choose their Python flavor.

@DerThorsten / @IsabelParedes / @wolfv / @jezdez / @ntoll maybe we should meet again and try to make progress on this?

**Comment by ntoll** [comment]
Just thinking out loud here... if Emscriptenforge have their own Python interpreter with their unique capabilities and features, why can't that be another Python interpreter available via PyScript..? Clearly I need to understand the technical aspects of what's there. But like I said, just a thought. @DerThorsten / @IsabelParedes / @wolfv..?

**Comment by ntoll** [comment]
Polite ping. Any movement on this?

**Comment by petercorke** [comment]
In May 2026 I demonstrated RTB working with WASM in the browser, at a tutorial that Tobi and I ran.  This is now mainstream in RTB, you can click on the "try it now" button on the README.md to see how it works.  My Machine Vision Toolbox similarly now works in the browser.

My current bugbear is JSPI.  Safari, my preferred browser, doesn't support it so I'm currently pinned to an old pyodide version.  Hopefully Apple WebKit gets with the program, or I need to ditch Safari in favour of Chrome.

**Comment by chrisbuerginrogers** [comment]
Worked like a charm - thanks - that is really cool.
____________________________________
ph: 617 627 2882  web<https://engineering.tufts.edu/me/people/faculty/chris-rogers>, cal<https://fantastical.app/crogers-bpB2/meetings-w-chris>, and zoom<https://tufts.zoom.us/my/crogers>

On Sep 14, 2026, at 10:13, Peter Corke ***@***.***> wrote:



In May 2026 I demonstrated RTB working with WASM in the browser, at a tutorial that Tobi and I ran. This is now mainstream in RTB, you can click on the "try it now" button on the README.md to see how it works. My Machine Vision Toolbox similarly now works in the browser.

My current bugbear is JSPI. Safari, my preferred browser, doesn't support it so I'm currently pinned to an old pyodide version. Hopefully Apple WebKit gets with the program, or I need to ditch Safari in favour of Chrome.
